### PR TITLE
Fix CopyObject regression calculating md5sum

### LINKED ISSUE #1044: make fails due to unit test failure in TestCoreCopyObject
```
--- FAIL: TestCoreCopyObject (1.06s)
	core_test.go:433: Error: expected etag to be same as source object 7515e2ca2f78be0c2437f3fb674412c6, but found different etag :d41d8cd98f00b204e9800998ecf8427e
=== RUN   TestCoreCopyObjectPart

```

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
CopyObject() failed to calculate proper md5sum
when without encryption headers. This is a regression
fix perhaps introduced in commit 5f6d717b7a

## Motivation and Context
Fixes https://github.com/minio/minio-go/issues/1044

## Regression
Yes introduced in 5f6d717b7a

## How Has This Been Tested?
Using `minio-go` functional and unit tests.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by harshavardhana** [comment]
Tests are already present for this issue in minio-go unit tests, but not in mint functional tests.

File: cmd/object-handlers.go [hunk 1]
```diff
@@ -813,8 +813,10 @@ func (api objectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
 		writeErrorResponse(w, toAPIErrorCode(ctx, err), r.URL, guessIsBrowserReq(r))
 		return
 	}
+
 	rawReader := srcInfo.Reader
-	pReader := NewPutObjReader(srcInfo.Reader, srcInfo.Reader, nil)
+	pReader := NewPutObjReader(rawReader, nil, nil)
```
Comment (donatello) [inline cmd/object-handlers.go]: I think second arg also needs to be `rawReader` - but @poornas can confirm.

**Review by donatello — DISMISSED** [review]

File: cmd/object-handlers.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/object-handlers.go]: no it is not @donatello - there is no reason to provide second reader without encryption key.

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-tls.sh | :x: [more...](#0f468114-9c92-4cf5-b414-de2d3bbf02bc) |
<a name="0f468114-9c92-4cf5-b414-de2d3bbf02bc"></a>
### 6868-56f7984/mint-tls.sh.log:
```
Running with
SERVER_ENDPOINT:      192.168.39.129:9000
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         1
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp e553496f1a29:/mint/log /tmp/mint-logs'

(1/7) Running aws-sdk-go tests ... done in 1 seconds
(2/7) Running aws-sdk-java tests ... done in 30 seconds
(3/7) Running mc tests ... FAILED in 1 minutes and 4 seconds
{
  "name": "mc",
  "duration": "1787",
  "function": "test_copy_object_with_sse",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-folder /tmp/.mc-8617 --quiet --no-color cp --encrypt-key myminio/mc-test-bucket-30156/prefix=32byteslongsecretkeymustbegiven1 myminio/mc-test-bucket-30156/prefix/mc-test-object-32082 myminio/mc-test-bucket-30156/mc-test-object-32082  >>> \n`myminio/mc-test-bucket-30156/prefix/mc-test-object-32082` -> `myminio/mc-test-bucket-30156/mc-test-object-32082`\nmc: <ERROR> Failed to copy `https://192.168.39.129:9000/mc-test-bucket-30156/prefix/mc-test-object-32082`. You did not provide the number of bytes specified by the Content-Length HTTP header.\nmc: <ERROR> Session safely terminated. To resume session `mc session resume nnsNeoog`"
}

Executed 2 out of 7 tests successfully.
```

**Comment by harshavardhana** [comment]
`mc` functional tests seem to be work fine

```
~ make test
Installing golint
Installing gocyclo
Installing deadcode
Installing misspell
Installing ineffassign
Running vet
Running fmt
Running lint
Running cyclo
Running deadcode
Checking dependencies
Checking for project in GOPATH
Building minio binary to './mc'
Running unit tests
?   	github.com/minio/mc	[no test files]
ok  	github.com/minio/mc/cmd	(cached)
ok  	github.com/minio/mc/pkg/console	(cached)
ok  	github.com/minio/mc/pkg/hookreader	(cached)
ok  	github.com/minio/mc/pkg/httptracer	(cached)
ok  	github.com/minio/mc/pkg/ioutils	(cached)
ok  	github.com/minio/mc/pkg/probe	(cached)
Running functional tests
Running test_make_bucket()
Running test_make_bucket_error()
Running test_put_object()
Running test_put_object_error()
Running test_put_object_0byte()
Running test_put_object_with_storage_class()
Running test_put_object_with_storage_class_error()
Running test_put_object_multipart()
Running test_get_object()
Running test_get_object_multipart()
Running test_presigned_post_policy_error()
Running test_presigned_put_object()
Running test_presigned_get_object()
Running test_cat_object()
Running test_mirror_list_objects()
Running test_mirror_list_objects_storage_class()
Running test_find()
Running test_find_empty()
Running test_watch_object()
Running test_put_object_with_sse()
Running test_put_object_with_sse_error()
Running test_put_object_multipart_sse()
Running test_get_object_with_sse()
Running test_cat_object_with_sse()
Running test_cat_object_with_sse_error()
Running test_copy_object_with_sse()
Running test_copy_object_with_sse2()
Running test_sse_key_rotation()
Running test_mirror_with_sse()
Running test_rm_object_with_sse()
Running test_config_host_add()
Running test_config_host_add_error()
```

**Review by donatello — DISMISSED** [review]

**Review by poornas — COMMENTED** [review]
there's still an issue with copying from sse-s3 encrypted object to unencrypted even with this fix.

**Comment by harshavardhana** [comment]
> there's still an issue with copying from sse-s3 encrypted object to unencrypted even with this fix.

This fix is not addressing that @poornas - it's only fixing what I could reproduce.

**Comment by poornas** [comment]
>     there's still an issue with copying from sse-s3 encrypted object to unencrypted even with this fix.
> 
> This fix is not addressing that @poornas - it's only fixing what I could reproduce.
what I meant was, this fix is causing copy to fail for this case whereas it copied object  on master branch (though with wrong etag)

**Comment by harshavardhana** [comment]
Can you point to a test case ? I am using all mc tests right now.

**Comment by poornas** [comment]
``` 
aws s3api put-object --bucket tbucket11 --key sses3  --body /home/kris/Downloads/smallfile --endpoint-url https://localhost:9000 --no-verify-ssl --server-side-encryption AES256

aws s3api copy-object --bucket tbucket11 --copy-source tbucket11/sses3 --key s3cpy   --endpoint-url https://localhost:9000 --no-verify-ssl

```
This fails with ```An error occurred (IncompleteBody) when calling the CopyObject operation: You did not provide the number of bytes specified by the Content-Length HTTP header.``` , source side didn't decrypt correctly

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-worm.sh | :heavy_check_mark: |
| mint-tls.sh | :x: [more...](#ba4a1e79-bb84-49d6-903d-22c5ef80cb57) |
| mint-compression-xl.sh | :x: [more...](#4eb352ea-df45-418b-93fa-bb64f7099808) |
| mint-xl.sh | :x: [more...](#998f77ae-ab63-4ae2-b737-c862f79afa0a) |
| mint-gateway-nas.sh | :x: [more...](#a3bde497-c308-40b3-92f7-8d0f0bd97a68) |
| mint-large-bucket.sh | :x: [more...](#28c4bb64-cd9e-4e72-8360-1c1b55458aab) |
| mint-compression-fs.sh | :x: [more...](#c2b897b3-d9ef-4d79-aeea-75ce2b343eeb) |
| mint-fs.sh | :x: [more...](#e751d22f-eac0-474e-b288-48828dd82eac) |
| mint-dist-xl.sh | :x: [more...](#a414438a-6832-4dd3-b23d-3fab2033cf7d) |
<a name="a414438a-6832-4dd3-b23d-3fab2033cf7d"></a>
### 6868-8963a9e/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30324
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp a0dbb7339f44:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 18 seconds
{
  "name": "awscli",
  "duration": 4235,
  "function": "aws --endpoint-url http://minikube:30324 s3api get-object --bucket awscli-mint-test-bucket-18159 --key awscli-mint-test-bucket-18159-object /tmp/datafile-0-b\n",
  "status": "FAIL",
  "error": "Etag mismatch for multipart 0 byte object"
}

Executed 0 out of 13 tests successfully.
```
<a name="28c4bb64-cd9e-4e72-8360-1c1b55458aab"></a>
### 6868-8963a9e/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30081
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp dcee6ee07f9f:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 22 seconds
{
  "name": "awscli",
  "duration": 4142,
  "function": "aws --endpoint-url http://minikube:30081 s3api get-object --bucket awscli-mint-test-bucket-31728 --key awscli-mint-test-bucket-31728-object /tmp/datafile-0-b\n",
  "status": "FAIL",
  "error": "Etag mismatch for multipart 0 byte object"
}

Executed 0 out of 13 tests successfully.
```
<a name="c2b897b3-d9ef-4d79-aeea-75ce2b343eeb"></a>
### 6868-8963a9e/mint-compression-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:31946
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp ac5a0dd5d847:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 1 minutes and 3 seconds
{
  "name": "awscli",
  "duration": 14971,
  "function": "aws --endpoint-url http://minikube:31946 s3api get-object --bucket awscli-mint-test-bucket-27797 --key awscli-mint-test-bucket-27797-object /tmp/datafile-0-b\n",
  "status": "FAIL",
  "error": "Etag mismatch for multipart 0 byte object"
}

Executed 0 out of 13 tests successfully.
```
<a name="ba4a1e79-bb84-49d6-903d-22c5ef80cb57"></a>
### 6868-8963a9e/mint-tls.sh.log:
```
Running with
SERVER_ENDPOINT:      192.168.39.129:9000
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         1
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 32db98205f6a:/mint/log /tmp/mint-logs'

(1/7) Running aws-sdk-go tests ... done in 1 seconds
(2/7) Running aws-sdk-java tests ... done in 31 seconds
(3/7) Running mc tests ... FAILED in 29 seconds
{
  "name": "mc",
  "duration": "315",
  "function": "test_copy_object_with_sse",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-folder /tmp/.mc-7974 --quiet --no-color cp --encrypt-key myminio/mc-test-bucket-9129/prefix=32byteslongsecretkeymustbegiven1 myminio/mc-test-bucket-9129/prefix/mc-test-object-7472 myminio/mc-test-bucket-9129/mc-test-object-7472  >>> \n`myminio/mc-test-bucket-9129/prefix/mc-test-object-7472` -> `myminio/mc-test-bucket-9129/mc-test-object-7472`\nmc: <ERROR> Failed to copy `https://192.168.39.129:9000/mc-test-bucket-9129/prefix/mc-test-object-7472`. You did not provide the number of bytes specified by the Content-Length HTTP header.\nmc: <ERROR> Session safely terminated. To resume session `mc session resume pmvUtcag`"
}

Executed 2 out of 7 tests successfully.
```
<a name="4eb352ea-df45-418b-93fa-bb64f7099808"></a>
### 6868-8963a9e/mint-compression-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:32734
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 150e967777b1:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 1 minutes and 4 seconds
{
  "name": "awscli",
  "duration": 15027,
  "function": "aws --endpoint-url http://minikube:32734 s3api get-object --bucket awscli-mint-test-bucket-2717 --key awscli-mint-test-bucket-2717-object /tmp/datafile-0-b\n",
  "status": "FAIL",
  "error": "Etag mismatch for multipart 0 byte object"
}

Executed 0 out of 13 tests successfully.
```
<a name="998f77ae-ab63-4ae2-b737-c862f79afa0a"></a>
### 6868-8963a9e/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:32079
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp bd9d110465f8:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 1 minutes and 3 seconds
{
  "name": "awscli",
  "duration": 14735,
  "function": "aws --endpoint-url http://minikube:32079 s3api get-object --bucket awscli-mint-test-bucket-15534 --key awscli-mint-test-bucket-15534-object /tmp/datafile-0-b\n",
  "status": "FAIL",
  "error": "Etag mismatch for multipart 0 byte object"
}

Executed 0 out of 13 tests successfully.
```
<a name="e751d22f-eac0-474e-b288-48828dd82eac"></a>
### 6868-8963a9e/mint-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30188
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp e10a76aa91e0:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 1 minutes and 2 seconds
{
  "name": "awscli",
  "duration": 14184,
  "function": "aws --endpoint-url http://minikube:30188 s3api get-object --bucket awscli-mint-test-bucket-1766 --key awscli-mint-test-bucket-1766-object /tmp/datafile-0-b\n",
  "status": "FAIL",
  "error": "Etag mismatch for multipart 0 byte object"
}

Executed 0 out of 13 tests successfully.
```
<a name="a3bde497-c308-40b3-92f7-8d0f0bd97a68"></a>
### 6868-8963a9e/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30056
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 32f945e7731e:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 1 minutes and 4 seconds
{
  "name": "awscli",
  "duration": 15029,
  "function": "aws --endpoint-url http://minikube:30056 s3api get-object --bucket awscli-mint-test-bucket-24278 --key awscli-mint-test-bucket-24278-object /tmp/datafile-0-b\n",
  "status": "FAIL",
  "error": "Etag mismatch for multipart 0 byte object"
}

Executed 0 out of 13 tests successfully.
```

**Comment by harshavardhana** [comment]
> ```
> aws s3api put-object --bucket tbucket11 --key sses3  --body /home/kris/Downloads/smallfile --endpoint-url https://localhost:9000 --no-verify-ssl --server-side-encryption AES256
> 
> aws s3api copy-object --bucket tbucket11 --copy-source tbucket11/sses3 --key s3cpy   --endpoint-url https://localhost:9000 --no-verify-ssl
> ```
> This fails with `An error occurred (IncompleteBody) when calling the CopyObject operation: You did not provide the number of bytes specified by the Content-Length HTTP header.` , source side didn't decrypt correctly

@poornas should be fixed in latest push

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by poornas** [comment]
tested - looks good. Can you rerun travis for this build? failure seems to be unrelated to this fix

**Review by poornas — APPROVED** [review]
LGTM

**Review by donatello — APPROVED** [review]

**Pull request merged**